dotnet/WPF/VisualLayerIntegration/VisualLayerIntegration/BarGraphUtility/BarGraph.cs

```diff
@@ -0,0 +1,479 @@
+        {
+            Single = 0,
+            Random = 1,
+            PerBarLinearGradient = 3,
+            AmbientAnimatingPerBarLinearGradient = 4
+        }
+
+        // Constructor for bar graph.
+        // To insert graph, call the constructor then use barGraph.Root to get the container to parent.
+        public BarGraph(Compositor compositor, IntPtr hwnd, string title, string xAxisLabel, 
+            string yAxisLabel, float width, float height, double dpiX, double dpiY, float[] data, WindowRenderTarget renderTarget, 
+            bool AnimationsOn = true, GraphBarStyle graphBarStyle = GraphBarStyle.Single,
+            List<Windows.UI.Color> barColors = null)
+        {
+            this.compositor = compositor;
+            this.hwnd = hwnd;
+            this.graphWidth = (float)(width * dpiX / 96.0);
+            this.graphHeight = (float)(height * dpiY / 96.0);
+
+            this.graphData = data;
+
+            Title = title;
+            XAxisLabel = xAxisLabel;
+            YAxisLabel = yAxisLabel;
+
+            this.graphBarStyle = graphBarStyle;
+
+            if (barColors != null)
+            {
+                graphBarColors = barColors;
+            }
+            else
+            {
+                graphBarColors = new List<Windows.UI.Color>() { Colors.Blue };
+            }
+
+            // Configure options for text.
+            var factory2D = new SharpDX.Direct2D1.Factory();
+
+            var properties = new HwndRenderTargetProperties();
+            properties.Hwnd = this.hwnd;
+            properties.PixelSize = new SharpDX.Size2((int)(width * dpiX / 96.0), (int)(width * dpiY / 96.0));
+            properties.PresentOptions = PresentOptions.None;
+
+            textRenderTarget = renderTarget;
+
+            // Generate graph structure.
+            var graphRoot = GenerateGraphStructure();
+            GraphRoot = graphRoot;
+
+            BarRoot = this.compositor.CreateContainerVisual();
+            GraphRoot.Children.InsertAtBottom(BarRoot);
+
+            // If data has been provided, initialize bars and animations; otherwise, leave graph empty.
+            if (graphData.Length > 0)
+            {
+                barValueMap = new Hashtable();
+                var bars = CreateBars(graphData);
+                AddBarsToTree(bars);
+            }
+        }
+
+        private void UpdateSizeAndPositions()
+        {
+            shapeGraphOffsetY = graphHeight * 1 / 15;
+            shapeGraphOffsetX = graphWidth * 1 / 15;
+            shapeGraphContainerHeight = graphHeight - shapeGraphOffsetY * 2;
+            shapeGraphContainerWidth = graphWidth - shapeGraphOffsetX * 2;
+
+            mainContainer.Offset = new System.Numerics.Vector3(shapeGraphOffsetX, shapeGraphOffsetY, 0);
+
+            barWidth = ComputeBarWidth();
+            barSpacing = (float)(0.5 * barWidth);
+
+            shapeContainer.Offset = new System.Numerics.Vector3(shapeGraphOffsetX, shapeGraphOffsetY, 0);
```

Comment:
Add a "using" rather than qualifying all these System.Numerics types.

Reply:
it ends up being an ambigious reference between system.numerics and sharpdx namespaces if i remove the inline qualification

Reply:
Time for aliases.

---
In reply to: [272758734](https://github.com/Microsoft/Windows.UI.Composition-Win32-Samples/pull/1#discussion_r272758734) [](ancestors = 272758734)